pa1/PPCodeUnitStream.h

```diff
@@ -0,0 +1,55 @@
+#ifndef PPCodeUnitStream_h
+#define PPCodeUnitStream_h
+
+#include "PPCodeUnit.h"
+#include "UTF32StreamIfc.h"
+#include "PPCodeUnitStreamIfc.h"
+#include <queue>
+#include <string>
+
+// Stream UTF32 code points into stream of PPCodeUnit that is easy to tokenize.
+//
+//   - Line splicing. i.e., occurrences of "\\\n" are replaced with
+//     PPCodeUnitType::WhitespaceCharacter so that we can revert it in raw strings.
+//
+//   - Comments are streamed into PPCodeUnitType::Comment objects. Other tools
+//     can extract the comment strings for more processing
+//
+//   - Digraphs and universal-character-names are also extracted out and can be
+//     reverted in raw strings.
+class PPCodeUnitStream: public PPCodeUnitStreamIfc {
+public:
+  PPCodeUnitStream(std::shared_ptr<UTF32StreamIfc>);
+
+  virtual bool isEmpty() const override;
+  virtual const std::shared_ptr<PPCodeUnit> getCodeUnit() const override;
+  virtual void toNext() override;
+
+  std::string getErrorMessage() const;
```

Comment:
Should document the following:

`PPCodeUnitStream` nevers throws exceptions, as with the rest of this project. Error reporting and handling is done on the caller site by explicitly checking the error message, as returned by `getErrorMessage()`. If the error message
- is empty, no error occurred in the last call to the `_pushCodeUnits()` function;
- is not empty, some error occurred there.

It is up to the caller site to print the error message, if any, in any way they want. They may choose to not print it at all.